File: newton_matrix.cpp
#include <stdio.h>
#include <math.h>
#include "der_diffusion_tensor.h"
#include "diffusion_tensor.h"
#include "newton_matrix.h"

void newton_matrix(Matrix<double> &u, double h, double a, Block<double> &D, Block<double> &L, Block<double> &U)
{
    int n = size(u, 2);
    int m = size(u, 1);

    double f = -a*0.5/(h*h);
    double z;    

    Block<double> B(n, m, m);
    Block<double> Bp(n, m, m);
    diffusion_tensor(u, B);

    for (int j = 1; j <= n; j++)
    {
        for (int i = 1; i <= m; i++)
        {
            for (int p = 1; p <= m; p++)
            {
                D(j, p, i)=0;
                L(j, p, i)=0;
                U(j, p, i)=0;
            }
        }
    }

    for (int r = 1; r <= m; r++)
    {
        der_diffusion_tensor(u, Bp, r);
        for (int p = 1; p <= m; p++)
        {
            for (int i = 2; i <= n; i++)
            {
                z = B(i, p, r) + B(i -1, p, r);
                D(i, p, r) = D(i, p, r) - f*z;
                L(i - 1, p, r) = L(i - 1, p, r) + f*z;
                
                for (int q = 1; q <= m; q++)
                    D(i, p, r) = D(i, p, r) - f*Bp(i, p, q)*(u(q, i)-u(q, i-1));

                for (int q = 1; q <= m; q++)
                    L(i - 1, p, r) = L(i - 1, p, r) - f*Bp(i - 1, p, q)*(u(q, i)-u(q, i-1));
            }
            for (int i = 1; i <= n - 1; i++)
            {
                z = B(i + 1, p, r) + B(i, p, r);
                U(i, p, r) = U(i, p, r) + f*z;
                D(i, p, r) = D(i, p, r) - f*z;
                
                for (int q = 1; q <= m; q++)
                    D(i, p, r) = D(i, p, r) + f*Bp(i, p, q)*(u(q, i+1)-u(q, i));
                
                for (int q = 1; q <= m; q++)
                    U(i, p, r) = U(i, p, r) + f*Bp(i + 1, p, q)*(u(q, i+1)-u(q, i));
            }
        }
    }

    for (int i = 1; i <= n; i++)
        for (int r = 1; r <= m; r++)
            D(i, r, r) = D(i, r, r) + 1;
}

File: newton_matrix.m
function [D, L, U]=newton_matrix(u, h, a, D, L, U)

%int:n,m,i,j,p,q,r
%double:a,h,f,z
%vector:
%matrix:B,Bp,u,D,L,U

n=size(u, 2);
m=size(u, 1);

f=-a*0.5/h^2;
B=zeros(m, m, n);
Bp=zeros(m, m, n);
B=diffusion_tensor(u, B);

for j=1:n
  for i=1:m
    for p=1:m
      D(p, i, j)=0;
      L(p, i, j)=0;
      U(p, i, j)=0;
    end
  end
end

for r=1:m
  Bp=der_diffusion_tensor(u, Bp, r);
  for p=1:m
    for i=2:n
      z=B(p, r, i)+B(p, r, i-1);
      D(p, r, i)=D(p, r, i)-f*z;
      L(p, r, i-1)=L(p, r, i-1)+f*z;
      for q=1:m
	D(p, r, i)=D(p, r, i)-f*Bp(p, q, i)*(u(q, i)-u(q, i-1));
      end
      for q=1:m
	L(p, r, i-1)=L(p, r, i-1)-f*Bp(p, q, i-1)*(u(q, i)-u(q, i-1));
      end
    end
    for i=1:n-1
      z=B(p, r, i+1)+B(p, r, i);
      U(p, r, i)=U(p, r, i)+f*z;
      D(p, r, i)=D(p, r, i)-f*z;
      for q=1:m
	D(p, r, i)=D(p, r, i)+f*Bp(p, q, i)*(u(q, i+1)-u(q, i));
      end
      for q=1:m
	U(p, r, i)=U(p, r, i)+f*Bp(p, q, i+1)*(u(q, i+1)-u(q, i));
      end

    end
  end
end
for i=1:n
  for r=1:m
    D(r, r, i)=D(r, r, i)+1;
  end
end